src/routing/lazy_messaging.rs

```diff
@@ -0,0 +1,353 @@
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::{
+    consensus::Vote,
+    error::Result,
+    messages::{Message, MessageHash, Variant},
+    network::Network,
+    node::Node,
+    section::Section,
+};
+use sn_messaging::DstLocation;
+use std::cmp::Ordering;
+use xor_name::XorName;
+
+/// On reception of an incoming message, determine the actions that need to be taken in order to
+/// bring ours and the senders knowledge about each other up to date.
+pub(crate) fn process(
+    node: &Node,
+    section: &Section,
+    network: &Network,
+    msg: &Message,
+) -> Result<Actions> {
+    let src_name = msg.src().name();
+    let src_prefix = if section.prefix().matches(&src_name) {
+        Some(section.prefix())
+    } else {
+        network
+            .section_by_name(&src_name)
+            .1
+            .map(|info| &info.prefix)
+    };
+    let src_key = msg.proof_chain_last_key();
+
+    let mut actions = Actions::default();
+    let mut send_other_section = false;
+
+    if let Some(src_prefix) = src_prefix {
+        if !src_prefix.matches(&node.name())
+            && !src_key
+                .map(|src_key| network.has_key(src_key))
+                .unwrap_or(false)
+        {
+            send_other_section = true;
+        }
+    }
+
+    if let Some(new) = msg.dst_key() {
+        if let Some(src_prefix) = src_prefix {
+            let old = network
+                .knowledge_by_name(&src_name)
+                .unwrap_or_else(|| section.chain().root_key());
+
+            if section.chain().cmp_by_position(new, old) == Ordering::Greater {
+                actions.vote.push(Vote::TheirKnowledge {
+                    prefix: *src_prefix,
+                    key: *new,
+                });
+            }
+        }
+
+        if new != section.chain().last_key() {
+            send_other_section = true;
+        }
+    }
+
+    if send_other_section {
+        // TODO: if src has split, consider sending to all child prefixes.
+        let dst_key = network.key_by_name(&src_name).cloned();
+        actions.send.push(create_other_section_message(
+            node,
+            section,
+            network,
+            src_name,
+            *msg.hash(),
+            dst_key,
+        )?);
+    }
+
+    Ok(actions)
+}
+
+fn create_other_section_message(
+    node: &Node,
+    section: &Section,
+    network: &Network,
+    dst: XorName,
+    nonce: MessageHash,
+    dst_key: Option<bls::PublicKey>,
+) -> Result<Message> {
+    let dst_knowledge = network
+        .knowledge_by_name(&dst)
+        .unwrap_or_else(|| section.chain().root_key());
+    let proof_chain = section
+        .chain()
+        .minimize(vec![section.chain().last_key(), dst_knowledge])?;
```

Comment:
shall the order of these two swapped ?
our own chain shall always being not shorter than other sections' knowledge of us?

Reply:
The order is not significant. `minimize` gives you the shortest sub-chain that contains all the given keys, regardless of in what order you pass them.